src/System.Text.Encoding/tests/UnicodeEncoding/UnicodeEncodingDecode.cs

```diff
@@ -63,66 +60,91 @@ public static IEnumerable<object[]> Decode_TestData()
+            EncodingHelpers.Decode(new UnicodeEncoding(true, false, false), bigEndianBytes, index, count, expected);
+            EncodingHelpers.Decode(new UnicodeEncoding(true, true, false), bigEndianBytes, index, count, expected);
+
+            // Decoding invalid bytes should throw with a DecoderExceptionFallback
+            NegativeEncodingTests.Decode_Invalid(new UnicodeEncoding(false, true, true), littleEndianBytes, index, count);
+            NegativeEncodingTests.Decode_Invalid(new UnicodeEncoding(false, false, true), littleEndianBytes, index, count);
+            NegativeEncodingTests.Decode_Invalid(new UnicodeEncoding(true, false, true), bigEndianBytes, index, count);
+            NegativeEncodingTests.Decode_Invalid(new UnicodeEncoding(true, true, true), bigEndianBytes, index, count);
         }
 
         [Fact]
-        public void Decode_InvalidUnicode()
+        public void Decode_InvalidBytes()
         {
-            // TODO: add into Decode_TestData once #7166 is fixed
+            // TODO: add into Decode_TestData or Decode_InvalidBytes_TestData once #7166 is fixed
             // High BMP non-chars
             Decode(new byte[] { 253, 255 }, 0, 2, "\uFFFD");
             Decode(new byte[] { 254, 255 }, 0, 2, "\uFFFE");
             Decode(new byte[] { 255, 255 }, 0, 2, "\uFFFF");
 
             // Invalid bytes
             byte[] validSurrogateBytes = new byte[] { 0, 216, 0, 220 };
-            Decode_LittleEndian(validSurrogateBytes, 0, 3, "\uFFFD\uFFFD");
-            Decode_LittleEndian(validSurrogateBytes, 1, 3, "\u00D8\uFFFD");
-            Decode(validSurrogateBytes, 0, 2, "\uFFFD");
-            Decode_LittleEndian(validSurrogateBytes, 1, 2, "\u00D8");
-            Decode(validSurrogateBytes, 2, 2, "\uFFFD");
-            Decode(validSurrogateBytes, 2, 1, "\uFFFD");
-
-            Decode(new byte[] { 97 }, 0, 1, "\uFFFD");
-            Decode(new byte[] { 97, 0, 97 }, 0, 3, "a\uFFFD");
+            Decode_InvalidBytes(validSurrogateBytes, 0, 3, "\uFFFD\uFFFD");
+            Decode_InvalidBytes(validSurrogateBytes, 1, 3, "\u00D8\uFFFD");
+            Decode_InvalidBytes(validSurrogateBytes, 0, 2, "\uFFFD");
+            Decode(validSurrogateBytes, 1, 2, "\u00D8");
+            Decode_InvalidBytes(validSurrogateBytes, 2, 2, "\uFFFD");
+            Decode_InvalidBytes(validSurrogateBytes, 2, 1, "\uFFFD");
+
+            Decode_InvalidBytes(new byte[] { 97 }, 0, 1, "\uFFFD");
+            Decode_InvalidBytes(new byte[] { 97, 0, 97 }, 0, 3, "a\uFFFD");
+
+            Decode_InvalidBytes(new byte[] { 3, 216, 48 }, 0, 3, "\uFFFD\uFFFD");
+        }
+
+        public byte[] GetBigEndianBytes(byte[] littleEndianBytes, int index, int count)
+        {
+            byte[] bytes = new byte[littleEndianBytes.Length];
+            for (int i = index; i < index + count; i += 2)
+            {
+                if (i + 1 >= bytes.Length)
```

Comment:
do we expect to get a case like this? if not, then I would just remove the check and make it assert instead.